main.c

```diff
@@ -143,36 +143,42 @@ enum WatchChildrenState {
  * \param stdinbuf Key presses to send to the auth child, if set.
  * \return If true, authentication was successful and the program should exit.
  */
-int WatchChildren(Display *dpy, Window w, enum WatchChildrenState state,
-                  const char *stdinbuf) {
-  int auth_running = WantAuthChild(state == WATCH_CHILDREN_FORCE_AUTH);
+int WatchChildren(Display *dpy, Window auth_win, Window saver_win,
+                  enum WatchChildrenState state, const char *stdinbuf) {
+  int want_auth = WantAuthChild(state == WATCH_CHILDREN_FORCE_AUTH);
+  int auth_running;
 
-  // Note: auth_running is true whenever we WANT to run authentication, or it is
+  // Note: want_auth is true whenever we WANT to run authentication, or it is
   // already running. It may have recently terminated, which we will notice
   // later.
-  if (auth_running) {
-    // Make sure the saver is shut down to not interfere with the screen.
-    WatchSaverChild(dpy, w, 0, saver_executable, 0);
-
+  if (want_auth) {
     // Actually start the auth child, or notice termination.
-    if (WatchAuthChild(dpy, w, auth_executable,
+    if (WatchAuthChild(dpy, auth_win, auth_executable,
                        state == WATCH_CHILDREN_FORCE_AUTH, stdinbuf,
                        &auth_running)) {
       // Auth performed successfully. Terminate the other children.
-      WatchSaverChild(dpy, w, 0, saver_executable, 0);
+      WatchSaverChild(dpy, saver_win, 0, saver_executable, 0);
       // Now terminate the screen lock.
       return 1;
     }
   }
 
   // No auth child is running, either because we don't have any running and
   // didn't start one, or because it just terminated.
-  // Show the screen saver.
   if (!auth_running) {
-    WatchSaverChild(dpy, w, 0, saver_executable,
-                    state != WATCH_CHILDREN_SAVER_DISABLED);
+    XUnmapWindow(dpy, auth_win);
```

Comment:
Can we unmap only right after exit? Otherwise we'd be calling XUnmapWindow every frame which seems wasteful (but harmless).

Could be done e.g. by just moving this block into the next if block.

Reply:
I intentionally kept it outside to avoid the corner case of leaving the auth window mapped somehow, but playing around with it some more it seems moving it into the next if block is reliable enough.